scalding-serialization-macros/src/main/scala/com/twitter/scalding/serialization/macros/impl/ordered_serialization/SealedTraitLike.scala

```diff
@@ -0,0 +1,318 @@
+ */
+package com.twitter.scalding.serialization.macros.impl.ordered_serialization
+
+import scala.language.experimental.macros
+import scala.reflect.macros.Context
+
+import com.twitter.scalding._
+import com.twitter.scalding.serialization.OrderedSerialization
+
+object SealedTraitLike {
+
+  /**
+   * Compare Binary for generating similar types of binary comparasion code
+   * Args:
+   * inputStreamA: should contain the variable name that has the input stream A bound to
+   * inputStreamB: should contain the variable name that has the input stream B bound to
+   * subData: Its a list of the sub components of this sealed trait, for each one
+   *          we include an index of this sub type, the clase class/type of this sub type,
+   *          and finally a means to compare two instances of this type.
+   */
+  def compareBinary(c: Context)(inputStreamA: c.TermName, inputStreamB: c.TermName)(subData: List[(Int, c.Type, TreeOrderedBuf[c.type])]): c.Tree = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+    val valueA = freshT("valueA")
+    val valueB = freshT("valueB")
+    val idxCmp = freshT("idxCmp")
+
+    val compareSameTypes: Tree = subData.foldLeft(Option.empty[Tree]) {
+      case (existing, (idx, tpe, tBuf)) =>
+
+        val commonCmp: Tree = tBuf.compareBinary(inputStreamA, inputStreamB)
+
+        existing match {
+          case Some(t) =>
+            Some(q"""
+              if($valueA == $idx) {
+                $commonCmp
+              } else {
+                $t
+              }
+            """)
+          case None =>
+            Some(q"""
+                if($valueA == $idx) {
+                  $commonCmp
+                } else {
+                  sys.error("unreachable code -- this could only be reached by corruption in serialization.")
+                }""")
+        }
+    }.get
+
+    q"""
+        val $valueA: Int = $inputStreamA.readByte.toInt
+        val $valueB: Int = $inputStreamB.readByte.toInt
+        val $idxCmp: Int = _root_.java.lang.Integer.compare($valueA, $valueB)
+        if($idxCmp != 0) {
+          // Since compare matches, only look at valueA from now on.
+          $idxCmp
+        } else {
+          $compareSameTypes
+        }
+      """
+  }
+
+  def hash(c: Context)(element: c.TermName)(subData: List[(Int, c.Type, TreeOrderedBuf[c.type])]): c.Tree = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    subData.foldLeft(Option.empty[Tree]) {
+      case (optiExisting, (idx, tpe, tBuf)) =>
+        val innerArg = freshT("innerArg")
+        val elementHash: Tree = q"""
+              val $innerArg: $tpe = $element.asInstanceOf[$tpe]
+              ${tBuf.hash(innerArg)}
+            """
+
+        optiExisting match {
+          case Some(s) =>
+            Some(q"""
+            if($element.isInstanceOf[$tpe]) {
+              $elementHash ^ _root_.com.twitter.scalding.serialization.Hasher.int.hash($idx)
```

Comment:
But your points valid, i'm not sure we need to eval or do anything special here. We can probably just call the method itself